### PR TITLE
Coap fixes

### LINKED ISSUE #1601: coap client command causing segmentaion fault
~~I have had a look around and I believe that this could possibly be due to the recent changed to the COAP API in commit b6e72560a9cf73dd47a7d6a36fc7c827bfd7d91e.~~

When running the `coap client ...` command in the POSIX application it will crash.

I have run it through gdb and this was the output:
```
(gdb) run
Starting program: /home/user/Documents/openthread/output/i686-pc-linux-gnu/bin/ot-cli-ftd 1
> ifconfig up
Done
> ipaddr
fdde:ad00:beef:0:9832:4707:d09c:9a7d
fe80:0:0:0:48ed:5826:6782:8615
Done
> coap server start
Server started with resource '': Done
> coap client put fdde:ad00:beef:0:9832:4707:d09c:9a7d test con Message
                                                                       Program received signal SIGSEGV, Segmentation fault.
                                                                                                                           Thread::Coap::CoapBase::NewMessage (this=0x80cfffc <sInstanceRaw+13948>, aHeader=..., aPriority=2 '\002')
    at /home/user/Documents/openthread/examples/../src/core/coap/coap_base.cpp:45
45	    VerifyOrExit(aHeader.GetLength() >= Header::kMinHeaderLength);
(gdb) where
#0  Thread::Coap::CoapBase::NewMessage (this=0x80cfffc <sInstanceRaw+13948>, aHeader=..., aPriority=2 '\002')
    at /home/user/Documents/openthread/examples/../src/core/coap/coap_base.cpp:45
#1  0x0806af57 in otCoapNewMessage (aInstance=0x80cc980 <sInstanceRaw>, aHeader=0x0) at /home/user/Documents/openthread/examples/../src/core/api/coap_api.cpp:131
#2  0x080503d9 in Thread::Cli::Coap::ProcessClient (argc=5, argv=0xbfffecc0) at /home/user/Documents/openthread/examples/../src/cli/cli_coap.cpp:259
...
```

Edit: I have since done a checkout of the initial commit of the coap client/server and it didn't work back then either, the change to the COAP API is therefore not the issue here

### PR DESCRIPTION
**Pull request by CampbellWray**
Resolves #1601 and creates adds check for NULL header in the API to avoid a similar issue in the future

### CODE REVIEW

**Comment by CampbellWray** [comment]
I signed it!

File: src/core/api/coap_api.cpp [hunk 1]
```diff
@@ -128,7 +128,13 @@ const otCoapOption *otCoapHeaderGetNextOption(otCoapHeader *aHeader)
 
 otMessage *otCoapNewMessage(otInstance *aInstance, const otCoapHeader *aHeader)
 {
-    return aInstance->mThreadNetif.GetCoapClient().NewMessage(*(static_cast<const Coap::Header *>(aHeader)));
+    Message *message;
+
```
Comment (jwhui) [inline src/core/api/coap_api.cpp]: nit: remove trailing whitespace.

File: src/core/api/coap_api.cpp [hunk 2]
```diff
@@ -128,7 +128,13 @@ const otCoapOption *otCoapHeaderGetNextOption(otCoapHeader *aHeader)
 
 otMessage *otCoapNewMessage(otInstance *aInstance, const otCoapHeader *aHeader)
 {
-    return aInstance->mThreadNetif.GetCoapClient().NewMessage(*(static_cast<const Coap::Header *>(aHeader)));
+    Message *message;
+    
+    VerifyOrExit(aHeader != NULL, message = NULL);
+    message = aInstance->mThreadNetif.GetCoapClient().NewMessage(*(static_cast<const Coap::Header *>(aHeader)));
+
```
Comment (jwhui) [inline src/core/api/coap_api.cpp]: nit: remote trailing whitespace.

**Comment by jwhui** [comment]
Thanks for submitting this!  Overall, looks good.  The code-style check is failing.  To correct that, you'll want to:

1. Rebase on the latest openthread master branch.
2. Remove extra extra whitespace as noted in my comments.

**Comment by CampbellWray** [comment]
I think I did that right, I guess we'll see

**Comment by CampbellWray** [comment]
I may have done that in the most stupid way possible, my rebase failed the first time but I think its right now

**Comment by CampbellWray** [comment]
Cheers

**Pull request merged**